### PR TITLE
Mpi4py

### PR DESCRIPTION
**Pull request by allaffa**

### CODE REVIEW

File: utils/utils.py [hunk 1]
```diff
@@ -61,6 +76,10 @@ def setup_ddp():
     return distributed_data_parallelism, world_size, world_rank
 
 
+def kill_mpi_environment():
```
Comment (streeve) [inline utils/utils.py:79]: I would just call directly

File: utils/utils.py [hunk 2]
```diff
@@ -38,17 +48,22 @@ def setup_ddp():
...
     try:
+        """
         world_size = os.environ["OMPI_COMM_WORLD_SIZE"]
         world_rank = os.environ["OMPI_COMM_WORLD_RANK"]
-        os.environ["MASTER_ADDR"] = master_addr
-        os.environ["MASTER_PORT"] = master_port
         os.environ["WORLD_SIZE"] = world_size
         os.environ["RANK"] = world_rank
+        """
+        os.environ["MASTER_ADDR"] = master_addr
```
Comment (streeve) [inline utils/utils.py:63]: Is there no way to do this with mpi4py?

File: utils/utils.py [hunk 3]
```diff
@@ -18,13 +18,23 @@
...
+mpi4py.rc.initialize = False
+mpi4py.rc.finalize = False
+from mpi4py import MPI
+
 
 def get_comm_size_and_rank():
     world_size = 1
     world_rank = 0
     try:
+        """
```
Comment (streeve) [inline utils/utils.py:32]: Remove

File: utils/utils.py [hunk 4]
```diff
@@ -38,17 +48,22 @@ def setup_ddp():
...
         world_size = os.environ["OMPI_COMM_WORLD_SIZE"]
         world_rank = os.environ["OMPI_COMM_WORLD_RANK"]
-        os.environ["MASTER_ADDR"] = master_addr
-        os.environ["MASTER_PORT"] = master_port
         os.environ["WORLD_SIZE"] = world_size
         os.environ["RANK"] = world_rank
+        """
+        os.environ["MASTER_ADDR"] = master_addr
+        os.environ["MASTER_PORT"] = master_port
+        world_size = MPI.COMM_WORLD.Get_size()
```
Comment (streeve) [inline utils/utils.py:65]: Call `get_comm_size_and_rank`

File: utils/utils.py [hunk 5]
```diff
@@ -38,17 +48,22 @@ def setup_ddp():
     backend = "nccl" if torch.cuda.is_available() else "gloo"
 
     distributed_data_parallelism = False
+    MPI.Init()
```
Comment (streeve) [inline utils/utils.py:51]: Probably needs to be guarded with `MPI.Is_initialized()`

File: utils/utils.py [hunk 6]
```diff
@@ -61,6 +76,10 @@ def setup_ddp():
     return distributed_data_parallelism, world_size, world_rank
 
 
+def kill_mpi_environment():
+    MPI.Finalize()
```
Comment (streeve) [inline utils/utils.py:80]: Not sure the right way to ensure other processes continue in the session and only the last one calls

**Review by streeve — CHANGES_REQUESTED** [review]

**Review by jychoi-hpc — APPROVED** [review]
It looks ok to me.

**Comment by streeve** [comment]
Closed in favor of #42

**Pull request closed without merging**